Level 3

MDKirianova opened the pull request:
Готова функциональность на реакте, можно зарегистрироваться, авторизоваться, внести изменения в профиль. Авторизованным пользователям доступны страницы со всеми видео и сохраненными видео. Можно сохранять и удалять видео из избранного. На главной странице есть 2 шапки - для авторизованных и для неавторизованных пользователей

### Примечания к последней версии сайта на level-3

IP 84.201.165.247

Frontend https://mdkirianova.bitfilms.nomoredomainsmonster.ru

Backend https://api.mdkirianova.bitfilms.nomoredomainsmonster.ru

MDKirianova committed:
level-3 test

MDKirianova committed:
level-3 movies api and search first commit

MDKirianova committed:
level-3 movies api and search second commit

MDKirianova committed:
level-3 login and register validation, token check, add and remove movies

MDKirianova committed:
level-3 profile and saved movies

MDKirianova committed:
level-3 backend errors and register fix

MDKirianova committed:
level-3 saved movies search fix

MDKirianova committed:
level-3 fix routing

MDKirianova committed:
level-3 fix movies request

MDKirianova committed:
level-3 add movie api error

MDKirianova committed:
level-3 fix api link

MDKirianova committed:
level-3 fix comments and readme

MDKirianova committed:
level-3 fix app

MDKirianova committed:
level-3 fix app

MDKirianova committed:
level-3 fix popup

MDKirianova committed:
level-3 fix preloader

MDKirianova committed:
level-3 fix readme

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: Сохранять данные  поиска на странице «Сохранённые фильмы» в localstorage не нужно. Сейчас эти данные заносятся в localstorage в те же переменные search-query и is-checkbox-checked, что и данные со страницы фильмов, заменяя их значения, что ведёт к неправильному повторному показу результата поиска на странице фильмов. Нужно, чтобы при входе на страницу «Сохранённые фильмы» пользователь видел все свои избранные карточки без задания параметров поиска.**
            
    **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить строка поиска и статус чекбокса «Короткометражки» - логическое значение, показыающее в активном ("короткометражки") или дефолтном положении находится чекбокс. При появлении пользователя на странице фильмов вновь после того, как он перезагрузит сайт, или перейдёт на страницу фильмов с другой страницы, сохранённую в локальном хранилище информацию нужно будет использовать для вывода на экран результата поиска, который был перед перезагрузкой или уходом на другую страницу (с расстановкой актуального состояния лайков). При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать.** 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
	**Комментарий: При вводе в поле email адреса почты без точки перед именем домена верхнего уровня (в правой части адреса, например "review@gmailcom") можно выполнить запрос на сервер, что противоречит текущему требованию чек-листа, так как в этом случае адрес будет невалидным. Нужно, чтобы валидация имени, почты и пароля и на серверной и на клиентской частях приложения, была одинаковой  по строгости, и при невалидных данных запрос на сервер вообще не возможно было отправить.**

   **Комментарий: Еесли на клиенте происходит недостаточно строгая валидация адреса email, и на сервер можно отослать невалидный адрес с точки зрения сервера, нужно подобрать регулярное выражение, рекомендуемое в интернете для проверки email, чтобы строгость проверки email на клиенте, совпадала со строгостью проверки email на сервере. Тогда при запросах на регистрацию или авторизацию, сервер не будет присылать в ответе ошибку 400, признавая, что адрес почты валиден. В этом и состоит смысл текущего пункта чек-листа.**
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    **Комментарий: На странице профиля при попытке пользователя сохранить адрес почты, который принадлежит другому пользователю и сервер приложения присылает ошибку 409 в ответе,  всё равно выдаётся сообщение об успешном сохранении профиля (см. снимок).**
![Снимок экрана в 2024-02-13 00-52-16](https://github.com/MDKirianova/movies-explorer-frontend/assets/19558891/96333467-ba98-48f9-a74d-1df3129e456d)
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
	  **Комментарий: При выводе результата поиска при передвижении чекбокса, без нажатия кнопки "Поиск" на форме поиска, должен работать тот же алгоритм, который работает при сабмите формы поиска, то есть работает при нажатии кнопки "Поиск". Таким образом, при передвижении чекбокса для получения результата поиска должен производиться новый фильтр всего массива фильмов по заданной в текущий момент строке поиска в форме поиска и положению самого чекбокса. Сейчас при передвижении чекбокса для вывода результата берётся строка поиска, которая была при предыдущем поиске, а не та, которая находится в текущий момент в форме поиска, что в определённых случаях приводит к показу неверного результата (см. снимок). Сделайте в соответствии с этим замечанием и алгоритм поиска на странице «Сохранённые фильмы».**
![Снимок экрана в 2024-02-13 00-23-59](https://github.com/MDKirianova/movies-explorer-frontend/assets/19558891/ac685c7d-018f-461f-b71b-4b3da496305a)	  
	**Комментарий: данные о результате поиска со страницы фильмов должны сохраняться в localstorage при каждом поиске, независимо от того каким способом произведён поиск, то есть и при выводе результата поиска от нажатия кнопки "Поиск", и при выводе результата поиска вследствие передвижения чекбокса. Сейчас новая строка поиска при передвижении чекбокса в другое положение не сохраняется.**  
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  **Комментарий: Вывод карточек на экран не соответствует постановке задачи, данной в описании диплома. Нужно протестировать и отладить работу функционала в соответствии с приведёнными ниже требованиями и подсказками при различных действиях пользователя и разной ширине экрана.**

**Ориентироваться нужно на описание функционала рендера карточек, которое дано в задании по диплому:**
   Обратите внимание, что количество карточек, которые отображаются на странице, зависит от ширины экрана устройства.
    • Ширина 1280px — 4 ряда карточек. Кнопка «Ещё» загружает дополнительный ряд карточек.
    • Ширина 768px — 4 ряда карточек. Кнопка «Ещё» загружает ряд карточек.
    • Ширина от 320px до 480px — 5 карточек по 1 в ряд. Кнопка «Ещё» загружает по 2 карточки.
В ряду отображайте столько карточек, сколько отображается в вашем макете на соответствующей ширине экрана.
  
  **Комментарий (пояснение, как должен быть реализован функционал вывода карточек, чтобы удовлетворить требованиям, приведённым в цитате выше):**
    
  **1. При изменении пользователем ширины экрана результат поиска нужно выводить с самого начала с применением дебаунса. То есть, когда пользователь остановится на какой-то ширине экрана, результат нужно выводить снова, начиная с первых 4-х или 5-ти рядов, после которых вывести кнопку "Ещё". Результат выводить в соответствии с правилами для текущей ширины экрана.** 

  **2. При осуществлении нового поиска на той же ширине экрана, при которой был предыдущий поиск, вывод карточек нужно начинать с самого начала схемы вывода карточек, то есть с первых 4-х (или 5-ти) рядов, после которых нужно вывести кнопку "Ещё".**
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
    **Комментарий: при установке и снятии активного статуса кнопки лайка на карточках, которые были выведены на экран с помощью кнопки «Ещё», отображение этих карточек исчезает с экрана, и появляется кнопка «Ещё». Это очень не удобно для пользователя, так как карточка, на которой устанавливался или снимался лайк, становится не видной, и до неё опять нужно пролистывать ряды карточек. Никаких дополнительных событий при установке и снятии активного статуса кнопки лайка происходить не должно.**
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [ ] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  **Комментарий: При попытке перейти на любой защищённый роут происходит редирект на страницу авторизации, а не на `/` .**
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: По URL-адресу авторизованный пользователь может зайти и на страницы авторизации и регистрации, что не допустимо, так как открывается возможность авторизоваться ещё раз, не выйдя из предыдущего аккаунта.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: Все исходные числовые данные проекта - длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще", другие постоянные числовые данные, значения которым задаются до начала выполнения приложения, нужно выносить в отдельный файл констант, чтобы эти исходные данные было легко менять. В коде принято использовать только имена этих констант, чтобы не было "магических" значений.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 56.
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет.**

MDKirianova committed:
level-3 fix popup on profile page, if error popup is correct

MDKirianova committed:
level-3 add email pattern for register, login, profile pages

MDKirianova committed:
level-3 fix searchform and chekbox for saved-movies and movies/ localstorage works only for movies

MDKirianova committed:
level-3 onSubmit on search form,
the input value appears in the request and in localstorage both by button and checkbox

MDKirianova committed:
level-3 fix cards grid, add bouncer and hook useWindowSize on movie page

MDKirianova committed:
level-3 fix navigate in protected route - to /

MDKirianova committed:
level-3 fix favorite click, remove  movies rendering on array change

MDKirianova committed:
level-3 change baseUrl of MainApi to .env

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.  
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.  
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: По URL-адресу авторизованный пользователь может зайти и на страницы авторизации и регистрации, что не допустимо, так как открывается возможность авторизоваться ещё раз, не выйдя из предыдущего аккаунта.**
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: Все исходные числовые данные проекта - длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще", другие постоянные числовые данные, значения которым задаются до начала выполнения приложения, нужно выносить в отдельный файл констант, чтобы эти исходные данные было легко менять. В коде принято использовать только имена этих констант, чтобы не было "магических" значений.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [ ] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
**Комментарий: при входе на страницы «Регистрация» и «Авторизация» на них присутствуют дефолтные данные. Пользователю приходится их стирать, чтобы ввести свои. На незаполненых полях этих форм должны быть плейсхолдеры, дефолтные данные использоваться не должны.**
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 93. 
## У Вас очень хороший дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Основы TypeScript
https://code-basics.com/ru/languages/typescript

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Борис Черный
Профессиональный TypeScript. Разработка масштабируемых JavaScript-приложений

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**
